pysingfel/detector/epix10k.py

```diff
@@ -1,16 +1,118 @@
+            self.switchPointVariations = [2.*8, 0.] ## keV, placeholder
+            self.switchPointWidths = [2.*8, 0.] ## keV, placeholder
+            self.offsets = [0, 1100] ## ADU; placeholder
+            self.gains = [12.5, 0.1] ## ADU/keV
+            self.gainErrors = [0.01, 0.015] ## dimensionless, placeholder
+            self.nonLinearity = [0.01/800, 0.017/self.maxEnergy] ## 1/keV, placeholder
+
+        self.minE = [0.] + self.switchPoints[0:self.nRanges-1] ## keV
+        ## this is for adding nonlinearity
+        ## calculated with respect to start of range
+        self.nonLinearity = np.array(self.nonLinearity)
+        self.nonLinearity *= 0. ## comment out to get nonlinearity
+
+        self.setupMatrices()
+        ## the idea is to allow us to use a scalar to describe the 2d array
+        ## corresponding to a particular characteristic
+        ## or to smear a 2d array
+        ## or one might use a numpy array to set the 2d array
+
+    def updateFlatGains(self, flatGains):  ## not sure this makes sense in the residual world
+        self.gains = self.matricize(flatGains)
+        raise Exception
+
+    def updateFlatOffsets(self, flatOffsets):
+        self.offsets = self.matricize(flatOffsets)
 
-class Epix10kDetector(LCLSDetector):
-    def __init__(self, *args, **kwargs):
-        super(Epix10kDetector, self).__init__(*args, **kwargs)
+    def updateFlatSwitchPoints(self, flatSwitchPoints):
+        self.switchPoints = self.matricize(flatSwitchPoints)
+
+    def updateSwitchPoints(self, switchPoints):
+        self.switchPoints = switchPoints
+
+    def setupMatrices(self):
+        self.residualGains = self.matricize(self.gains, self.gainErrors, None)/self.matricize(self.gains, None, None) # Q: ADU/keV?
+        self.offsets = self.matricize(self.offsets)
+        self.switchPoints = self.matricize(self.switchPoints, None, self.switchPointVariations)
+
+    def matricize(self, array, relativeSmear=None, absoluteSmear=None): # better to separate out smearing function
+        base = np.ones((self.panel_num, self.panel_pixel_num_x[0], self.panel_pixel_num_y[0]))
+        tmp = []
+        [tmp.append(array[i]*base) for i in range(self.nRanges)]
+        if relativeSmear is not None: ## should probably be by range, handle array or scalar
+            if not np.isscalar(relativeSmear):
+                print("temp check in relativeSmear array handler")
+                if len(relativeSmear)==self.nRanges:
+                    for n in range(self.nRanges): ## lazy and unidiomatic
+                        smears = 1 + (np.random.normal(size=self.panel_num*self.panel_pixel_num_x[0]*self.panel_pixel_num_y[0]).reshape((self.panel_num, self.panel_pixel_num_x[0], self.panel_pixel_num_y[0]))).clip(-3,3)*relativeSmear[n]
+                        ## clip to eliminate unfortunate tails, e.g. negative gain 
+                        tmp[n] *= smears
+                else:
+                    raise Exception
```

Comment:
Please add some description in the `Exception`.

Reply:
Added `raise Exception("nRanges not matched")` for now. (Perhaps could change to something better later)